---
title: "EWAS Analysis for NIMHAMS"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##Introduction

This script will  analyse the cerebellum. After completion of this script, the 'Comparison_to_BB_Cere' compares these results to the UK population

## Load the data
All the data and librares required for this analysis are to be loaded. 

```{r loading data, warning= FALSE, message=FALSE}
library(dplyr)
library(qqman)
library(gplots)
library(pheatmap)



setwd("/mnt/data1/NIMHANS/Down_stream_analysis/") ## change to working directory path
epicManifest<-read.csv("/mnt/data1/IndiaWorkshop/epicManifest.csv", stringsAsFactors = F, header=T)
load("NIMHAM_CER_Normalised_snpsremoved.rdat") 
```

## Run the EWAS on cerebellum

We have several co-variants to consider. In this experiment, we will consider the chip location, age and sex as we have already seperated the brain regions.

```{r linear model for frontal, message=FALSE, warning = FALSE}
#set the required columns to the correct type e.g character or numeric
pheno$Age <- as.numeric(as.character(pheno$Age))
pheno$Chip<- as.character(pheno$Chip)

#The commented section below requires only to be run once
# 
# #create a table which will be filled with the models coefficient values
# res<-matrix(data = NA, nrow = nrow(betas), ncol = 6)
# colnames(res)<-c("Age_Beta", "Age_SE", "Age_P", "Sex_Beta", "Sex_SE", "Sex_P")
# rownames(res)<-rownames(betas)
# for(i in 1:nrow(betas)){
#     model<-lm(betas[i,] ~ pheno$Age + factor(pheno$Sex)+factor(pheno$Chip))
#     res[i,1]<-coefficients(model)["pheno$Age" ]
#     res[i,2]<-summary(model)$coefficients["pheno$Age",2]
#     res[i,3]<-summary(model)$coefficients["pheno$Age",4]
#     res[i,4]<-coefficients(model)["factor(pheno$Sex)M" ]
#     res[i,5]<-summary(model)$coefficients["factor(pheno$Sex)M",2]
#     res[i,6]<-summary(model)$coefficients["factor(pheno$Sex)M",4]
#     }
# head(res)
# 
# #need to make the beta and SE as percentage
# res[,"Sex_Beta"] <- res[,"Sex_Beta"]*100
# res[,"Sex_SE"] <- res[,"Sex_SE"]*100
# res[,"Age_Beta"] <- res[,"Age_Beta"]*100
# res[,"Age_SE"] <- res[,"Age_SE"]*100

# write.csv(res, 'res_c_chip.csv')
```

Once the res table is generated. Save as it taken times to usuallygenerate the table

```{r annotations, message=FALSE, warning = FALSE}
# read the res file
# res <- read.csv('res_c_chip.csv', header = T, row.names = 1)
# epicManifest<-epicManifest[match(rownames(res), epicManifest$Name),]
# res<-cbind(res, epicManifest[,c("CHR", "MAPINFO", "UCSC_RefGene_Name", "UCSC_RefGene_Group", "Relation_to_UCSC_CpG_Island", "Regulatory_Feature_Name", "Regulatory_Feature_Group", "DNase_Hypersensitivity_NAME", "OpenChromatin_NAME", "TFBS_Evidence_Count", "Methyl450_Loci", "SNP_ID", "SNP_DISTANCE", "SNP_MinorAlleleFrequency")])
# 
# write.csv(res, "res_c_chip_included_manifest.csv") 

```


##Visualising the results

```{r QQplots, message=FALSE}
res <- read.csv("/mnt/data1/NIMHANS/Down_stream_analysis/NIMHAMSa/cereb/res_c_chip_included_manifest.csv", stringsAsFactors = F, header=T, row.names = 1)
par(mfrow = c(2,2))
qq(res$Sex_Beta)
qq(res$Age_Beta)
```


```{r manhatten plots, message=FALSE}

res<-res[which(res$CHR != "Y"),] ## to also exclude the X chromosome repeat and edit this line
res<-res[which(res$CHR != "X"),]
res$CHR<-as.character(res$CHR)
res$CHR<-as.numeric(res$CHR)
res<-res[which(res$CHR != ""),]
bonfP<-0.05/nrow(res)



manhattan(res, p = "Age_P", bp = "MAPINFO", chr = "CHR", genomewide = -log10(bonfP), suggestiveline = -log10(5e-5), logp=T, col=c("blue","yellow"), ylim=c(0,8))
manhattan(res, p = "Sex_P", bp = "MAPINFO", chr = "CHR", genomewide = -log10(bonfP), suggestiveline = -log10(5e-5), logp=T, col=c("blue","yellow"), ylim=c(0,8))

```


Scatter plot of significant robes.

```{r scatter plot , warning= F, message=FALSE}
#Order based on most to least significant probes
res<-res[order(res$Age_P),]
par(mfrow = c(2,3)) ## will plot 6 figures per page in 2 rows by 3 columns
for(i in 1:6){
  plot(pheno$Age, betas[rownames(res)[i],]*100 , xlab = "Age", ylab = "DNA methylation (%)", pch = 16, main = rownames(res)[i]) ## we multiple by 100 to plot of a % scale rather than proportion
  ## to add line of best fit
  model<-lm(betas[rownames(res)[i],]*100 ~ pheno$Age)
  abline(model, col = "red")  
}

#Order based on most to least significant probes
res<-res[order(res$Sex_P),]
par(mfrow = c(2,3)) ## will plot 6 figures per page in 2 rows by 3 columns
for(i in 1:6){
  plot(pheno$Sex, betas[rownames(res)[i],]*100 , xlab = "Sex", ylab = "DNA methylation (%)", pch = 16, main = rownames(res)[i]) 
  model<-lm(betas[rownames(res)[i],]*100 ~ pheno$Sex)
  abline(model, col = "red")  
}

```



Heatmap based on the order of most significant probes between sexes

```{r pheatmap sex, message=FALSE} 
#Order and select first 500 probes. Then takes these probes from betas matrix to get their raw read outs
res<-res[order(res$Sex_P),]
res_sig <- res[1:500,]
betas_r <- betas[rownames(res_sig),]
sigma<-apply(betas, 1, sd)

#This chunk of code generates the meta data so pheatmap knows what each probe is
cell_rows <- as.data.frame(pheno$Sex, rownames(pheno))
colnames(cell_rows)<- "Sex"
cell_age <- as.data.frame(pheno$Age, rownames(pheno))
colnames(cell_age)<- "Age"
cell_chip <- as.data.frame(pheno$Chip, rownames(pheno))
colnames(cell_chip)<- "Chip"
cell_rows <- cbind(cell_rows, cell_age, cell_chip)

pdf('/mnt/data1/NIMHANS/Down_stream_analysis/NIMHAMSa/cereb/pheatmaps_var_sex.pdf', 
    onefile = T)
pheatmap(betas_r, scale = 'row', 
         annotation_col = cell_rows,
         cluster_rows = F,
         show_rownames = F,
         show_colnames = F,
         main="Top significant probes between sexes")


pheatmap(betas[order(sigma, decreasing = TRUE)[1:500],], scale = 'row', 
         annotation_col = cell_rows,
         cluster_rows = F,
         show_rownames = F,
         show_colnames = F,
         main="Top varying probes")
dev.off()

``` 


```{r pheatmap of ages}
#Order and select first 500 probes. Then takes these probes from betas matrix to get their raw read outs
res<-res[order(res$Age_P),]
res_sig <- res[1:500,]
betas_r <- betas[rownames(res_sig),]

#This chunk of code generates the meta data so pheatmap knows what each probe is
cell_rows <- as.data.frame(pheno$Sex, rownames(pheno))
colnames(cell_rows)<- "Sex"
cell_age <- as.data.frame(pheno$Age, rownames(pheno))
colnames(cell_age)<- "Age"
cell_chip <- as.data.frame(pheno$Chip, rownames(pheno))
colnames(cell_chip)<- "Chip"
cell_rows <- cbind(cell_rows, cell_age, cell_chip)

pdf('/mnt/data1/NIMHANS/Down_stream_analysis/NIMHAMSa/cereb/pheatmaps_age.pdf')
pheatmap(betas_r, scale = 'row', 
         annotation_col = cell_rows,
         cluster_rows = F,
         show_rownames = F,
         show_colnames = F,
         main="Top significant probes between Ages")
dev.off()

```

